torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -272,6 +276,21 @@ def partition_parameters(self) -> List[List[Dict]]:
 
         return self._partition_parameters
 
+    def _update_trainable(self) -> None:
+        """Updates the partitioning and communication patterns if the trainability (`requires_grad`)
+        of some parameters changed.
+        """
+
+        # Create the optim which will work on the param shard
+        if not hasattr(self, "optim"):
+            self._clear_cache()
+            self._default_device = list(self.per_device_params.keys())[0]
+            self.optim = self._optim_constructor(self.partition_parameters()[self.rank], **self._optim_defaults)
+            self._sync_param_groups(self.optim.param_groups, self.param_groups)
+
+        if self.parameters_as_bucket_view:
+            self._setup_flat_buffers()
+
     @property
     def per_device_params(self) -> Dict[torch.device, List[List[Parameter]]]:
```

Comment:
do we expect users to access this for any reason? can this be private?

Reply:
ShardedDDP uses that typically, but it can do this and keep it private, this is python after all :)